### PR TITLE
Rule_L004 enforce indentation style from config

### LINKED ISSUE #562: Enhance L004 to enforce one indentation style across all files
Currently L004 throws an error if there is inconsistent indentation in a file (mixed spaces and tabs).

```
L: 121 | P:   1 | L004 | Mixed Tab and Space indentation found in file.
```

How would we feel about changing this to just enforce either tabs or spaces across all files? I personally would much prefer it as an anti-tabber, the logic would be simpler and a fix would be easy to implement. The indentation preference could just be set in the config.

### PR DESCRIPTION
**Pull request by NiallRees**
Resolves #562

This is a 'dumbing' down of the previous logic, but achieves a unified indentation style across all files. It will auto fix from tabs to spaces or vice versa, but only from spaces to tabs if the indent is a multiple of tab_space_size as explained in the code.

### CODE REVIEW

File: src/sqlfluff/core/rules/std.py [hunk 1]
```diff
@@ -714,25 +717,36 @@ class Rule_L004(BaseCrawler):
...
-                indents_union = indents_here | indents_seen
-                memory["indents_seen"] = indents_union
-                if len(indents_union) > 1:
-                    # We are seeing an indent we haven't seen before and we've seen others before
-                    return LintResult(anchor=segment, memory=memory)
-                else:
-                    return LintResult(memory=memory)
-        return LintResult(memory=memory)
+        correct_indent = (
+            " " * self.tab_space_size if self.indent_unit == "space" else "\t"
```
Comment (dmateusp) [inline src/sqlfluff/core/rules/std.py:729]: I think it'd look a little cleaner if you had 2 variables

`whitespace = " "`
and 
`tab = "\t"`

Reviewing the code it's not super natural to read right now ^^

**Review by dmateusp — COMMENTED** [review]
Just a small nit

**Comment by alanmcruickshank** [comment]
I like the thinking around this re-write - but it changes the intent of this rule a bit more. The original requires a newline and to recognize an indent, this new logic just looks for long patches of whitespace (regardless of whether they're at the start of a line).

What I'm not sure about is whether this is a bad thing. The potential issues would be:
- Handing indents (but these are only really possible with spaces *anyway* so probably not an issue).
- Long patches of whitespace *after* code, which it probably for layout reasons.

I think the former is probably fine, but I think the latter could behave strangely in cases like this:

```
SELECT
    a,         -- Some comment
    longer_col -- A lined up comment
FROM spam
```

In this example we have extended patches of whitespace *before and after* `a,`. The reason I'm not sure whether this is a problem is because I wonder if the same fixing logic actually applies fine 🤷 .

@NiallRees - could you add in a couple of test cases like this with a mixture of tabs and spaces in the before and after slots (and different configs) to make sure we don't behave strangely in these cases now that we're not looking for an explicit newline.

If it behaves nicely - then I think this is actually a **feature** rather than a bug and we should document it as such! 😄

**Comment by NiallRees** [comment]
> SELECT
>     a,         -- Some comment
>     longer_col -- A lined up comment
> FROM spam

Just done some testing with indents after text. I hadn't quite appreciated what text editors did with variable tab size to align comments. I think we're fine though for pre-text indents. This is the current result with indented comments after text:

image

I think the options are

1. Revert to previous in only trying to convert tabs to spaces which are directly after a newline
2. Introduce more complex text editor style understanding of tabs, and take into account the number of characters from the newline to determine how many space characters a tab should be converted to if it appears after other text.

I'm tempted to go with 1 for now as it's still an improvement over the previous rule (rule is triggered when 'wrong' indent used, and will convert from tabs->spaces and vice-versa), and is 'safe' in terms of not messing up formatting.

**Comment by alanmcruickshank** [comment]
I'm definitely fine with 1 for now. 2 could be fun, but I think it's a lot less important than the other. Using aligned comments after things is a lot less common than indentation in general (perhaps obviously 🤷 ).

File: test/core/rules/test_cases/L004.yml [hunk 2]
```diff
@@ -1,10 +1,106 @@
...
+  fail_str: |
+    SELECT
+    	a,			-- Some comment
+    	longer_col	-- A lined up comment
+    FROM spam
+
+  # The rule will only fix the indent before the select targets.
+  fix_str: |
+    SELECT
+        a,			-- Some comment
```
Comment (NiallRees) [inline test/core/rules/test_cases/L004.yml:83]: test demonstrating that the tab is replaced with spaces only after the newline, not after the text.

**Review by alanmcruickshank — APPROVED** [review]
Excellent test coverage!

**Pull request merged**